File: knapsack/cpp/knapsack.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

int 
KnapSack(int ks_weight, const std::vector<int>& iweight, const std::vector<int>& iprofit)
{
    std::vector<int> max_value(ks_weight + 1, 0);

    for (int i = 0; i < iweight.size(); ++i)
    {
        for (long long int w = ks_weight; w >= iweight[i]; --w) 
        {
            max_value[w] = std::max(max_value[w], max_value[w - iweight[i]] + iprofit[i]);
        }
    }
    
    return max_value[ks_weight];
}

int main()
{
    std::ios::sync_with_stdio(false);
    std::cin.tie(NULL);

    std::ifstream input_file("input.txt");

    if (!input_file) 
    {
        std::cerr << "Error opening file!" << "\n";
        return 1;
    }

    int item_number, total_weight;
    int section = 0;

    while (input_file >> item_number >> total_weight) 
    {
        std::vector<int> profit(item_number);
        std::vector<int> weight(item_number);

        for (int i = 0; i < item_number; ++i) 
        {
            int item_profit, item_weight;
            input_file >> item_profit >> item_weight;
            profit[i] = item_profit;
            weight[i] = item_weight;
        }

        std::cout << "Maximum profit for section " << section << ":" << "\n";
        std::cout << KnapSack(total_weight, weight, profit) << "\n";
        section++;
    }

    input_file.close();
    return 0;
}

File: knapsack/python/knapsack.py
def KnapSack(ks_weight, iweight, iprofit, inumber):
    dp = [0 for i in range(ks_weight + 1)]

    for i in range(1, inumber + 1):
        for w in range(ks_weight, 0, -1):
            if iweight[i - 1] <= w:
                dp[w] = max(dp[w], dp[w - iweight[i - 1]] + iprofit[i - 1])

    return dp[ks_weight]

if __name__ == '__main__':
    with open('knapsack_input.txt', 'r') as f:
        section = 0

        while True:
            item_number = f.readline()

            if item_number == "":
                break
            elif item_number == '\n':
                item_number = f.readline()
   
            item_number = int(item_number)
            total_weight = int(f.readline())
   
            item_profit = []
            item_weight = []
   
            for k in range(item_number):
                ip, iw = map(int, f.readline().split())

                item_profit.append(ip)
                item_weight.append(iw)
                        
            print(f"Maximum profit for section {section}:\n" + str(KnapSack(total_weight, item_weight, item_profit, item_number)))
            section += 1
